### PR TITLE
Scripts/Blackwing Lair - Doors correction

### LINKED ISSUE #24326: Scripts/Blackwing Lair: Doors not working anymore.
**Description:**

https://github.com/TrinityCore/TrinityCore/commit/99f31cbe5fc55e37bd4e00c77bfb35a2799c4ade#commitcomment-37879058

As of 99f31cbe5fc55e37bd4e00c77bfb35a2799c4ade script needs to be updated to cover correct doors and door behavior.

This issue is ignoring the template because its for being referenced by a PR.

**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

487370eb93b04fcda8ab2d31f22ea635db09a318

### PR DESCRIPTION
**Pull request by Jinnaix**
**Changes proposed:**

-  Fix all doors
-  Corrected the Chromaggus engage
-  Final door will be closed while fighting the boss and opened on reset/bosskill

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)

Closes #24326

**Tests performed:** (Does it build, tested in-game, etc.)
- [x] builds
- [x] tested in-game

**Known issues and TODO list:** (add/remove lines as needed)

- [x] need to rewrite the chromaggus lever

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp [hunk 1]
```diff
@@ -294,7 +297,44 @@ class boss_chromaggus : public CreatureScript
...
+
+            bool GossipHello(Player* player) override
+            {
+                if (instance->GetBossState(DATA_CHROMAGGUS) != DONE && instance->GetBossState(DATA_CHROMAGGUS) != IN_PROGRESS)
+                {
+                    instance->SetBossState(DATA_CHROMAGGUS, IN_PROGRESS);
+                    if (Creature* creature = instance->GetCreature(DATA_CHROMAGGUS))
+                        creature->AI()->JustEngagedWith(player);
+                    
+                    if (GameObject* gate = me->FindNearestGameObject(GO_CHROMAGGUS_DOOR, 65.0f))
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp]: we cant use the instance to find this gob? instead find near?

Comment (Ovahlord) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp]: *puts on obama voice*  yes we can!

Comment (Keader) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp]: hi sucker <3

Comment (Jinnaix) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp]: like this? 30f6e0006a42f769be33824542be4cd99c577c83

Comment (Ovahlord) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp]: No. You add the door to GameObjectData in the instance script and acces it via  instance->GetGameObject(type);

**Comment by Keader** [comment]
About https://github.com/TrinityCore/TrinityCore/pull/24327#discussion_r394465104
Is not incorrect, but is the "old way".
Need update like https://github.com/TrinityCore/TrinityCore/pull/24327#discussion_r394515101 suggested.
You can find examples in:

https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Outland/BlackTemple/instance_black_temple.cpp#L82-L88
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Outland/BlackTemple/instance_black_temple.cpp#L111 (you need call the generic hook)
And can use like: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Outland/BlackTemple/instance_black_temple.cpp#L163-L164 using DATA, instead guid

**Comment by Jinnaix** [comment]
Thank you @Keader 
I will look into this and "correct" the script afterwards.
Do we have a documentation about how to do things somewhere?

**Comment by Keader** [comment]
Nope.
Joschiwald implemented It years ago, has some documentation in commit message.
Any problem pm me in discord

**Comment by Jinnaix** [comment]
I didnt find any useable information on what should happen to the lever and the door to Chromaggus cage if you wipe. So once used, the door stays open. Will keep it that way until someone has valid info about that.

If everyone is happy now with current script state, im done here.

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp [hunk 2]
```diff
@@ -294,7 +298,45 @@ class boss_chromaggus : public CreatureScript
...
+class go_chromaggus_lever : public GameObjectScript
+{
+    public:
+        go_chromaggus_lever() : GameObjectScript("go_chromaggus_lever") { }
+
+        struct go_chromaggus_leverAI : public GameObjectAI
+        {
+            go_chromaggus_leverAI(GameObject* go) : GameObjectAI(go), instance(go->GetInstanceScript()) { }
+
+            InstanceScript* instance;
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp]: make this a private member, and PR will be ready to merge xD

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp [hunk 3]
```diff
@@ -327,6 +325,9 @@ class go_chromaggus_lever : public GameObjectScript
 
                 return true;
             }
+        
+        private:
+            InstanceScript* instance;
```
Comment (Ovahlord) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackwingLair/boss_chromaggus.cpp]: @Keader you sucker, it's _instance; It's a private member. Fix that and this should be good to go.

**Comment by Keader** [comment]
It's a minor correction, we can do while merging.
Anyway... i will merge when i get home

**Comment by Keader** [comment]
ty

**Pull request merged**